### ISSUE TITLE
NuGet drops file permissions when adding a file to the package on Unix

### ISSUE THREAD

**Issue by ViktorHofer** [NuGet/Home#13121]
### NuGet Product Used

dotnet.exe

### Product Version

7.0.9

### Worked before?

no

### Impact

It's more difficult to complete my work

### Repro Steps & Context
Ubuntu 22.04

```sh
dotnet new -n test
vi exec.sh
chmod +x exec.sh
vi test.csproj       // add: <ItemGroup><None Include="exec.sh" Pack="true" PackagePath="/" /></ItemGroup>
dotnet pack
sudo apt-get install unzip
unzip bin/Debug/test.1.0.0.nupkg -d a
ls -la a
```

Expected: 
```
-rwxr-xr-x 1 vihofer vihofer    0 Jan  2 17:40 exec.sh
```

Actual:
```
-rw-r--r-- 1 vihofer vihofer    0 Jan  2 16:40  exec.sh
```

**Comment by ViktorHofer** [comment]
NuGet Pack not handling this correctly makes it a non-starter to use a NuGet package to bundle executables or scripts on Unix into it.

**Comment by ViktorHofer** [comment]
From my conversation with copilot:

> Is the fundamental issue here that this can't be fixed because NuGet packages are just zips and preserving the executable bit isn't possible there in comparison to .tar.gz which handle this well or is it more about NuGet doing things in a wrong way?

It’s not fundamentally impossible because of `.nupkg` files
It’s not fundamentally impossible because .nupkg files are ZIPs. ZIP files can carry Unix permission bits, including the executable bit. The more accurate issue is that NuGet’s pack/restore/extraction pipeline does not currently treat Unix file modes as package metadata that must be preserved.

So this is closer to “NuGet is doing an incomplete/wrong thing for Unix packaging semantics” than “ZIP makes this impossible.”

Key points:
- .nupkg is a ZIP archive.
- ZIP entries have an “external attributes” field that can store Unix mode bits such as 0755.
- Tools like zip/unzip on Unix can preserve and restore executable bits when those attributes are set correctly.
- .tar.gz handles this more naturally because POSIX file mode metadata is part of the tar format’s core model.
- ZIP is more platform-ambiguous: permissions are optional/host-specific metadata, and not every ZIP producer/consumer preserves or applies them.

In your repro, the fact that:

>Text
-rwxr-xr-x exec.sh

becomes:

>Text
-rw-r--r-- exec.sh

strongly suggests that NuGet pack is not writing the Unix executable bit into the ZIP entry metadata, or NuGet’s ZIP-writing layer is normalizing/dropping it.

A proper fix would likely require NuGet to do both sides deliberately:

Pack: when running on Unix, read source file mode bits and encode them into the ZIP entry external attributes.
Restore/extract/install: when extracting on Unix, apply those mode bits back to the filesystem.
Spec/compat decision: define whether Unix modes are official NuGet package semantics or merely best-effort ZIP metadata.

So the short answer is:
ZIP is capable of preserving the executable bit. NuGet packages being ZIPs does not make this impossible. The issue is that NuGet currently does not preserve/apply Unix permission metadata in the way Unix users would expect. .tar.gz handles this more naturally, but NuGet could still support it with explicit ZIP metadata handling.

**Comment by ViktorHofer** [comment]
@jeffkl would your team accept a contribution in this area?

**Comment by jeffkl** [comment]
There are concerns in NuGet of adding functionality that would make it easier for people to run arbitrary scripts. We're trying to deprecate install.ps1 for packages.config in Visual Studio for example. 

A package can contain build logic which you could argue is arbitrary code that can be run. But at least that's quantifiable? One workaround is your package could contain build logic to mark the script as executable on non-Windows? That's still just running arbitrary code so 🤷 

But yeah the concern is mostly the security implications.

**Comment by ViktorHofer** [comment]
Right, as long as msbuild props and targets are read from a nuget package, arbitrary code execution is possible but explicitly documented as outside of our security guarantees (threat model). An msbuild target can just invoke chmod and fix the exec permissions (that's what some package do today) but that's a hack.

**Comment by baronfel** [comment]
This should be reopened - the lack of this makes .NET as an ecosystem look and feel materially worse on non-Windows platforms compared to our contemporaries. The executable bit itself isn't a security barrier (if the user owns the file/directory, they can _also_ change the permissions of the file!), it's purely a correctness/usability concern.

**Comment by lbussell** [comment]
I also hit this when trying to package scripts/templates inside a NuGet package (https://github.com/dotnet/docker-tools/issues/2181). Please reconsider fixing this bug - it would make packaging assets inside NuGet packages much simpler.
